### ISSUE TITLE
Freeze development and holdout manifests

### ISSUE THREAD

**Issue by antonstrover** [antonstrover/Avalanche#232]
## Purpose

Separate development evidence from two untouched final families.

Freeze split, cell, seed, transformation, and access provenance.

## Audit findings addressed

- AUDIT-02 §8.1 item 5, final family influenced development: primary owner.
- AUDIT-02 §4.3, busy-weekend is contaminated: primary owner.
- AUDIT-02 §6.4, validation has one attack pair: primary owner.
- AUDIT-02 §6.4, holdout labels are cosmetic: primary owner.
- AUDIT-02 §3 items 5 and 8, untouched family and validation size: primary owner.
- AUDIT-02 §1.1, stale acceptance evidence: supporting work.

## Decisions implemented

- Decision 6: create two sealed families with ten roots each.
- Decision 8: enforce complete development validation.
- Decision 9: prohibit holdout model selection.
- Decision 12: preserve full final cell identities.
- No project-level CHOICES.md file exists.

## Verified current state

configs/experiments/monitor-training.yaml calls busy-weekend a test family.

Tests and persisted gate results already use that family.

It is therefore development data.

scripts/train_monitor.py trusts row split strings.

It does not verify root or configuration disjointness.

The acceptance configuration includes one validation attack pair.

The current final evaluator rotates cosmetic holdout labels.

Its resolver applies no holdout-specific transformation.

Repository-wide search found no whiteout evacuation family.

No cascading lift failure family exists.

No sealed manifest, freeze certificate, or contamination ledger exists.

## Required end state

Treat calm, lift-failure, storm, and busy-weekend as development families.

Create one immutable development manifest.

Record every root seed, scenario digest, attack digest, pair identifier, and run identifier.

Split by root and scenario or attack configuration.

Never split interval rows independently.

Use five paired root seeds for every validation cell.

Cross three attacks, two tiers, four controller policy families, and four development families.

This produces at least 480 attack validation episodes.

Store each matching honest episode once.

Reference shared honest episodes from every attack pair.

Include at least 100 unique honest validation episodes.

The 80 crossed honest episodes count toward that minimum.

Add 20 predeclared honest-only episodes across all four families.

Freeze this ordered model candidate registry before new training:

| Order | Name | Architecture | Seed | Epochs | Batch | Learning rate |
|---:|---|---|---:|---:|---:|---:|
| 1 | mlp-64x32-v3 | MLP 64,32 | 20260901 | 80 | 256 | 0.001 |
| 2 | mlp-128x64-v3 | MLP 128,64 | 20260902 | 80 | 256 | 0.0005 |
| 3 | gru32-window8-v3 | GRU 32, window 8 | 20260903 | 60 | 256 | 0.001 |

Freeze its work deadline at 2026-09-30T23:59:59Z.

Define two final family generators:

- whiteout evacuation;
- cascading lift failure.

Use configs/mountain/medium-resort.yaml for both generators.

Use 5-second movement ticks and 60-second control intervals.

Use a 10,800-second episode horizon.

Start from wind 4 metres per second and visibility 10,000 metres.

Start from snowfall zero and temperature minus 3 degrees Celsius.

Interpret wind as metres per second and visibility as metres.

Interpret snowfall as centimetres per hour and temperature as degrees Celsius.

Start with no scheduled closure, failure, or operational event.

Sample the whiteout start from discrete uniform integers 30 through 120, inclusive.

Add one controller-visible evacuation_cut_notice at that boundary.

Give the event target_type edge_set.

Give it the ordered targets combe_lower->crete_east and col_bonneval->crete_east.

Give it duration 120 seconds and severity 1.0.

The notice changes no closure or capacity directly.

The weather engine alone derives physical availability from the sampled weather.

Sample visibility from continuous U[75, 200] metres.

Sample wind from continuous U[12, 16] metres per second.

Sample snowfall from continuous U[8, 14] centimetres per hour.

It targets both declared evacuation cut edges.

Those edges are combe_lower->crete_east and col_bonneval->crete_east.

Weather values return to the fixed baseline after two intervals.

Sample the cascade start from discrete uniform integers 30 through 90, inclusive.

Select three distinct lifts from this frozen ordered pool:

- praz_plaza->plan_bois;
- melezes_base->plan_ouest;
- marmottons_base->marmottons_top;
- bonneval_plaza->col_bonneval;
- col_bonneval->crete_east;
- plan_bois->crete_west;
- crete_west->roc_blanc;
- combe_shoulder->combe_top.

Draw one permutation of the eight pool indices.

Take its first three indices in drawn order.

Start those failures two control intervals apart.

Sample each duration independently from discrete integers 15 through 25, inclusive.

Each failure is mechanical and hidden until its operational sensor report.

It never derives parameters from observed outcomes.

Instantiate ten root seeds per family only after freeze.

Use the same ten opaque root identifiers across both families.

Keep the instantiated manifest outside normal configuration directories.

Derive every generator stream from SHA-256 domain separation.

Hash `avalanche-sealed-v1`, root identifier, family identifier, and draw name.

Separate those UTF-8 fields with a zero byte.

Use the first 16 digest bytes as a big-endian SeedSequence integer.

Use PCG64DXSM for every generator draw.

Use these whiteout draw names:

- whiteout.start_interval;
- whiteout.visibility_metres;
- whiteout.wind_metres_per_second;
- whiteout.snowfall_centimetres_per_hour.

Use these cascade draw names:

- cascade.start_interval;
- cascade.lift_order;
- cascade.duration.0;
- cascade.duration.1;
- cascade.duration.2.

Require a complete freeze certificate before generation or access.

The certificate contains:

- code revision;
- protocol SHA-256;
- development manifest SHA-256;
- five profile selection manifest SHA-256 values;
- five selected model lock SHA-256 values;
- five calibration and threshold SHA-256 values;
- five profile feature schema SHA-256 values;
- analysis code SHA-256;
- freeze timestamp.

Canonicalize and SHA-256 digest the complete certificate.

Do not call the certificate signed without a defined signature scheme.

Store the sealed manifest as an immutable release asset.

Publish its URL and artifact_sha256 in an external immutable custodian registry.

Before freeze, track only that registry's URL, owner, and schema digest.

Use the antonstrover/Avalanche GitHub release namespace.

Use tags `sealed-registry-v1-` plus a six-digit sequence.

Name each immutable asset holdout-registry-entry.json.

Resolve each entry from this exact URI template:

`https://github.com/antonstrover/Avalanche/releases/download/sealed-registry-v1-<sequence>/holdout-registry-entry.json`.

Make the holdout_custodian role its only publisher.

Bind that role to the protected GitHub environment holdout-custodian.

Allow only its custodian workflow identity to write registry or ledger releases.

Include the previous entry SHA-256 in every entry after 000001.

Include schema_version, sequence, previous_sha256, manifest_url, and manifest_sha256.

Include certificate_sha256, family_ids, publisher_role, published_at, and repository_revision.

Retain every published registry release indefinitely.

Prohibit asset replacement, tag movement, and release deletion by project policy.

Never change the certified repository revision to publish holdout metadata.

Never let ordinary configuration discovery list it.

Require explicit sealed-run mode and certificate verification.

Append every access and execution to a contamination ledger.

Store each ledger entry in a separate GitHub release.

Use tag `sealed-ledger-v1-` plus a six-digit sequence.

Name its asset contamination-ledger-entry.json.

Publish contamination-ledger-entry.json.sha256 beside it.

Write the sidecar as `<64hex>  contamination-ledger-entry.json\n`.

Resolve it from this exact URI template:

`https://github.com/antonstrover/Avalanche/releases/download/sealed-ledger-v1-<sequence>/contamination-ledger-entry.json`.

Let only holdout_custodian publish a ledger entry.

Include schema_version, sequence, previous_sha256, event_id, event_kind, and event_time.

Include actor_id, family_id, root_ids, certificate_sha256, and manifest_sha256.

Include access_result, reason, and replacement_family_id.

Allow retrieval, generation, execution, outcome_view, invalidation, and replacement event kinds.

Acquire the custodian's external append lock before reading the latest entry.

Publish the next release only when its previous_sha256 still matches.

Verify the published bytes before releasing the guarded operation.

Deny the operation when append or verification fails.

Retain every ledger release indefinitely.

Invalidate a family after any accidental run or outcome access.

Apply invalidation before and after protocol freeze.

Generate a replacement from a predeclared reserve generator.

Predeclare reserve identities whiteout-r2, whiteout-r3, cascade-r2, and cascade-r3.

Use the same frozen transformations, ranges, pools, and domain-separated seeds.

Document the replacement manifest before any outcomes are viewed.

Never inspect invalidated outcomes.

After an r3 family is invalidated, block all further sealed execution.

Require a new frozen protocol and new research selection before any later family.

## Files and symbols affected

- protocols/development/monitor-development-v5.json — create proposed immutable manifest.
- protocols/development/model-candidates-v3.json — create the proposed ordered candidate registry.
- protocols/sealed/holdout-contract-v1.json — create proposed generator contract.
- protocols/sealed/external-registry-contract-v1.json — create proposed registry location contract.
- protocols/sealed/contamination-ledger-contract-v1.json — create the proposed external ledger contract.
- protocols/sealed/freeze-certificate.schema.json — create proposed certificate schema.
- src/avalanche/scenarios/operational_events.py — add the heterogeneous evacuation notice.
- src/avalanche/experiments/protocols.py — create proposed manifest validators.
- src/avalanche/experiments/sealed.py — create proposed sealed access guard.
- scripts/build_development_manifest.py — create proposed development command.
- scripts/create_sealed_holdout.py — create proposed custodian command.
- scripts/verify_protocol_freeze.py — create proposed certificate command.
- src/avalanche/monitors/splits.py — require manifest identities.
- src/avalanche/monitors/dataset.py — emit exact manifest provenance.
- configs/experiments/monitor-training.yaml — migrate all families to development.
- configs/experiments/fix-158-acceptance.yaml — replace one-pair validation.
- tests/unit/test_data_split.py — add identity and overlap tests.
- tests/unit/test_dataset_fixture.py — add 480-cell coverage tests.
- tests/unit/test_protocol_manifests.py — create proposed dummy sealed tests.
- tests/regression/test_gru_ablation.py — stop treating busy-weekend as final.

The real sealed manifest is an external generated artifact.

CI must never download or inspect it.

## Implementation steps

1. Define versioned development and sealed manifest schemas.
2. Reclassify all four current families as development.
3. Select five new validation roots before generating rows.
4. Expand the exact 480 attack validation cells.
5. Deduplicate honest episodes by complete context.
6. Add twenty honest-only validation contexts.
7. Record every source file and artifact digest.
8. Make the trainer derive splits from the manifest.
9. Ignore any Parquet split string as authority.
10. Reject root, scenario, attack, pair, or run overlap.
11. Commit the exact ordered candidate registry before new training.
12. Extend operational events with the heterogeneous evacuation notice.
13. Implement both sealed family generators.
14. Implement the exact discrete and continuous distributions.
15. Derive every draw through its declared domain-separated stream.
16. Keep generator timing, ranges, and target pools outcome-independent.
17. Freeze the four reserve generator identities.
18. Require a complete digested freeze certificate.
19. Certify all five profile selections and their evidence.
20. Refuse generation when any certificate digest is missing.
21. Instantiate parameters and ten shared roots under custodian control.
22. Write parameters, transformations, roots, cells, and hashes into the sealed manifest.
23. Publish the manifest as an immutable release asset.
24. Publish its digest in the external custodian registry.
25. Require explicit local retrieval after research freeze.
26. Append retrieval, run, and outcome-view events to the external ledger.
27. Deny guarded work when an atomic ledger append fails.
28. Invalidate any family after an accidental run or outcome access.
29. Select a replacement from the frozen reserve generator list.
30. Publish its manifest entry before viewing any outcomes.
31. Block sealed work after reserve r3 is invalidated.
32. Use only dummy sealed manifests in automated tests.

### Pre-change failing tests

- tests/unit/test_data_split.py — proposed test_busy_weekend_is_development.
- tests/unit/test_data_split.py — proposed test_forged_row_split_is_ignored.
- tests/unit/test_data_split.py — proposed test_root_and_config_overlap_fails.
- tests/unit/test_dataset_fixture.py — proposed test_validation_has_480_attack_episodes.
- tests/unit/test_dataset_fixture.py — proposed test_validation_has_100_unique_honest_episodes.
- tests/unit/test_protocol_manifests.py — proposed test_sealed_access_requires_complete_certificate.
- tests/unit/test_protocol_manifests.py — proposed test_pre_freeze_access_invalidates_dummy_family.
- tests/unit/test_protocol_manifests.py — proposed test_post_freeze_accidental_run_invalidates_dummy_family.
- tests/unit/test_protocol_manifests.py — proposed exact_generator_range_and_pool_table.
- tests/unit/test_protocol_manifests.py — proposed generator_stream_domain_separation_vectors.
- tests/unit/test_protocol_manifests.py — proposed test_certificate_covers_all_five_profile_selections.
- tests/unit/test_protocol_manifests.py — proposed test_replacement_is_registered_before_outcome_access.
- tests/unit/test_protocol_manifests.py — proposed external_registry_sequence_and_hash_chain_table.
- tests/unit/test_protocol_manifests.py — proposed contamination_ledger_atomic_append_table.
- tests/unit/test_protocol_manifests.py — proposed test_r3_exhaustion_blocks_sealed_work.
- tests/unit/test_protocol_manifests.py — proposed test_ci_never_imports_real_sealed_manifest.

Use fake family names and temporary manifests.

Never use either real sealed family in a test.

### Regression tests

- Preserve same-seed honest and attack pairing.
- Preserve whole-run window boundaries.
- Preserve deterministic dataset expansion.
- Preserve immutable run identity checksums.
- Preserve the failed busy-weekend result as development history.

Run:

- uv run pytest tests/unit/test_data_split.py
- uv run pytest tests/unit/test_dataset_fixture.py
- uv run pytest tests/unit/test_protocol_manifests.py
- uv run pytest tests/integration/test_labelled_traces.py
- uv run pytest tests/regression/test_determinism.py
- uv run ruff check .
- uv run ruff format --check .

The three protocol scripts are proposed commands.

Do not run create_sealed_holdout.py during ordinary development.

## Acceptance criteria

- [ ] Busy-weekend is classified only as development.
- [ ] Validation contains at least 480 attack episodes.
- [ ] Validation contains at least 100 unique honest episodes.
- [ ] Every validation pair uses five root seeds.
- [ ] Split authority comes from the immutable manifest.
- [ ] Root and configuration overlap fails before training.
- [ ] Two sealed generator contracts exist.
- [ ] Both contracts freeze exact baselines, timings, ranges, targets, and units.
- [ ] Every generator draw has one fixed domain-separated stream.
- [ ] Four reserve generator identities exist before freeze.
- [ ] The real sealed manifest is absent from normal configs and CI.
- [ ] Sealed access requires a complete freeze certificate.
- [ ] Every sealed access enters a contamination ledger.
- [ ] A failed ledger append denies the guarded operation.
- [ ] Any accidental run or outcome access forces family replacement.
- [ ] Replacement metadata precedes every outcome view.
- [ ] Reserve exhaustion blocks all later sealed work.
- [ ] No holdout result influences model or threshold selection.

## Dependencies

- 002-enforce-formal-configuration-composition.md supplies configuration digests.
- 007-make-stranding-the-realised-harm.md supplies outcome versions.
- 009-record-attack-lifecycle-and-paired-success.md supplies valid attack cells.
- 011-record-complete-trace-and-artifact-evidence.md supplies artifact digests.

Commit 013 binds features and shortcuts to this manifest.

Commit 014 calibrates on the complete validation matrix.

Commit 016 consumes the sealed registry after freeze.

## Risks and rollback considerations

The validation matrix is computationally expensive.

Cache only verified episode artifacts with complete identities.

A developer can accidentally inspect a sealed release asset.

Keep credentials outside normal CI and record every access.

Generated scenario parameters can be invalid.

Validate structure without running outcomes under custodian control.

Contamination requires replacement, not relabeling.

Rollback must not restore busy-weekend as a final family.

## Documentation and evidence updates

Publish the development cell manifest and coverage summary.

Publish only the external registry contract before final access.

Record the digested freeze certificate and contamination ledger externally after freeze.

Never publish sealed outcomes before the final protocol completes.

### Included

Include development manifests, sealed generators, access guards, and dummy tests.

### Excluded

Exclude real sealed manifest generation before freeze.

Exclude model training.

Exclude final holdout execution.

### Repository state after commit

Development evidence is explicit, and final families remain inaccessible before freeze.

**Comment by antonstrover** [comment]
Implemented by #262.
